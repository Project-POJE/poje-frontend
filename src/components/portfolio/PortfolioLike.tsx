import styled from 'styled-components';
import { AiTwotoneHeart } from 'react-icons/ai';
import usePortfolioLike from '@src/hooks/portfolio/usePortfolioLike';

export default function PortfolioLike() {
  const { data, handleLikeClick } = usePortfolioLike();

  if (!data) return <></>;

  return (
    <Container onClick={() => handleLikeClick()} isLike={data.likeStatus}>
      <AiTwotoneHeart />
      <p>{data.likeCount}</p>
    </Container>
  );
}

const Container = styled.div<{ isLike: boolean }>`
  position: fixed;
  bottom: 4rem;
  left: 2rem;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  font-size: 2rem;
  cursor: pointer;
  color: ${({ isLike }) => (isLike ? 'red' : 'black')};

  p {
    font-size: 1rem;
    transform: translateY(-5px);
  }
`;
