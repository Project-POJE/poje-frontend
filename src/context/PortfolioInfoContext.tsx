import CommonModal from '@src/components/common/CommonModal';
import { useGetNoteSender } from '@src/hooks/note';
import usePortfolioModifyPermission from '@src/hooks/portfolio/usePortfolioModifyPermission';
import { createContext, useMemo, useContext } from 'react';
import { useParams } from 'react-router-dom';

type ContextType = {
  portfolioId: string;
  permissionModify: boolean;
  portfolioEmail: string;
};
const PortfolioInfoContext = createContext<ContextType | undefined>(undefined);

type ContextPropType = {
  children: React.ReactNode;
};

type ParamsType = {
  portfolioId: string;
};

export default function PortfolioInfoContextProvider({
  children,
}: ContextPropType) {
  const { portfolioId } = useParams() as ParamsType;
  const permissionModify = usePortfolioModifyPermission(portfolioId);
  const portfolioEmail = useGetNoteSender(portfolioId);

  const context = useMemo<ContextType>(() => {
    return { portfolioId, permissionModify, portfolioEmail };
  }, [portfolioId, permissionModify, portfolioEmail]);

  return (
    <PortfolioInfoContext.Provider value={context}>
      <CommonModal />
      {children}
    </PortfolioInfoContext.Provider>
  );
}

export const usePortfolioInfo = () =>
  useContext(PortfolioInfoContext)! as ContextType;
