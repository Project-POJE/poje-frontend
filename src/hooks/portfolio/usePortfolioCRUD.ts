import useConfirmModal from '@src/hooks/common/useConfirmModal';
import { useNavigate } from 'react-router-dom';
import service from '@src/service';
import {
  CreatePortfolioTemplateRequest,
  DeletePortfolioItemRequest,
} from '@src/service/types/portfolio';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import useModal from '../common/useModal';

export default function usePortfolioCRUD() {
  const queryClient = useQueryClient();
  const { setModal } = useModal();
  const { setConfirmModal } = useConfirmModal();
  const navigate = useNavigate();
  const { mutate: createPortfolio } = useMutation(
    (data: CreatePortfolioTemplateRequest) =>
      service.portfolio.createPortfolioTemplate(data),
    {
      onSuccess: async (data) => {
        const {
          data: {
            result: { portfolioId },
          },
        } = data;
        queryClient.invalidateQueries(['portfolios']);
        queryClient.invalidateQueries(['portfoliosList']);
        queryClient.invalidateQueries(['myPortfolios']);
        setModal('포트폴리오 생성완료', () => {
          navigate(`/portfolio/${portfolioId}`);
        });
      },
      onError: (data: Error) => {
        console.log(data.message);
        setModal(data.message);
      },
    }
  );

  const { mutate: deletePortfolio } = useMutation(
    ({ portfolioId }: DeletePortfolioItemRequest) =>
      service.portfolio.deletePortfolio({ portfolioId }),
    {
      onSuccess: () => {
        queryClient.invalidateQueries(['portfolios']);
        queryClient.invalidateQueries(['portfoliosList']);
        queryClient.invalidateQueries(['myPortfolios']);
      },
    }
  );

  const onDeletePortfolio = (portfolioId: string) => {
    setConfirmModal('정말로 삭제 하시겠습니까?', () => {
      deletePortfolio({ portfolioId });
    });
  };

  return {
    createPortfolio,
    onDeletePortfolio,
  };
}
