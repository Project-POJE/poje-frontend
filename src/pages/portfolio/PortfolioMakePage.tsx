export default function PortfolioMakePage() {
  return <div>ProtfolioMakePage</div>;
}
