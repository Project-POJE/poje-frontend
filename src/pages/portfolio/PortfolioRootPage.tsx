import { Outlet } from 'react-router-dom';
import PortfolioHeader from '../../components/portfolio/PortfolioHeader';

export default function PortfolioRootPage() {
  return (
    <>
      <PortfolioHeader />
      <Outlet />
    </>
  );
}
