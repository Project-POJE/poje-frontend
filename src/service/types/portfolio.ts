import { BasicResponse } from './basicResponse';

export interface createPortfolioTemplateRequest {
  job: string;
}

export interface createPortfolioTemplateResponse extends BasicResponse {}
