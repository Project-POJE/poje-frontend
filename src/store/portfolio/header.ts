import { atom } from 'recoil';

const isDisapperIntro = atom<boolean>({
  key: 'isDisapperIntro',
  default: false,
});
