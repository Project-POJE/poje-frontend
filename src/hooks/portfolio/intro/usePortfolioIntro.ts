import { useSetRecoilState } from 'recoil';
import service from '@src/service';
import { PortfolioIntroType } from '@src/service/types/portfolio';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { usePortfolioInfo } from '@src/context/PortfolioInfoContext';
import { useEffect, useLayoutEffect, useState } from 'react';
import { introLoading } from '@src/store/portfolio/loading';
import { isModifyModeFromPortfolioIntro } from '@src/store/portfolio/modify';
import { queryKey } from '@src/react-query/queryKey';

type CopiedPfIntroType = Omit<
  PortfolioIntroType,
  'likeStatus' | 'likeCount'
> & {
  backgroundImgFile: File | null;
};

export default function usePortfolioIntro() {
  const queryClient = useQueryClient();
  const [copiedPfIntro, setCopiedPfIntro] = useState<CopiedPfIntroType>({
    title: '',
    description: '',
    portfolioId: '',
    jobName: '',
    backgroundImg: '',
    backgroundImgFile: null,
  });
  const setIsLoading = useSetRecoilState(introLoading);
  const setModify = useSetRecoilState(isModifyModeFromPortfolioIntro);
  const { portfolioId } = usePortfolioInfo();
  //todo 인트로 보내기 가져오기

  //1. suspense 사용시 onSuccess에 setState를 등록하면 업데이트가 안됨 이유는 unmount상태에서 변경했기때문
  const { data, isFetching } = useQuery(
    [queryKey.portfolioIntro, portfolioId],
    async () => await service.portfolio.getPortfolioIntro({ portfolioId }),
    {
      suspense: true,
    }
  );

  const updateIntro = useMutation<
    PortfolioIntroType,
    unknown,
    Pick<
      CopiedPfIntroType,
      'backgroundImgFile' | 'title' | 'description' | 'portfolioId'
    >,
    unknown
  >(
    ({ title, description, backgroundImgFile, portfolioId }) => {
      return service.portfolio.modifyPortfolioIntro({
        title,
        description,
        img: backgroundImgFile!,
        portfolioId,
      });
    },
    {
      onMutate: () => {
        setIsLoading(true);
        //queryClient.invalidateQueries(['portfolioIntro', portfolioId]);
      },
      onSuccess: (data) => {
        queryClient.setQueryData([queryKey.portfolioIntro, portfolioId], data);
      },
      onSettled: () => {
        setIsLoading(false);
        setModify(false);
      },
    }
  );

  //1에 대한 해결법으로 useLayoutEffect 걸어놓으면 됨 , 결과의 데이터를 기준으로
  useLayoutEffect(() => {
    data && setCopiedPfIntro((e) => ({ ...data, backgroundImgFile: null }));
  }, [data]);

  return {
    pfIntro: data!,
    updateIntro: updateIntro.mutate,
    copiedPfIntro,
    setCopiedPfIntro,
    isFetching,
  };
}
