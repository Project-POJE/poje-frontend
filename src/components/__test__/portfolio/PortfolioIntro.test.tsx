import PortfolioIntro from '@src/components/portfolio/introPart/PortfolioIntro';
import { usePortfolioInfo } from '@src/context/PortfolioInfoContext';
import usePortfolioIntro from '@src/hooks/portfolio/intro/usePortfolioIntro';
import { mockPortfolioIntro } from '@src/test_util/mock/data/portfolio';
import { withAllContext } from '@src/test_util/util';
import { render, screen } from '@testing-library/react';

jest.mock('@src/hooks/portfolio/intro/usePortfolioIntro');
jest.mock('@src/context/PortfolioInfoContext');

const mockUsePortfolioIntro: jest.Mock =
  usePortfolioIntro as unknown as jest.MockedFunction<any>;
const mockUsePortfolioInfo: jest.Mock =
  usePortfolioInfo as unknown as jest.MockedFunction<any>;

describe('PortfolioIntro Component', () => {
  afterEach(() => {
    mockUsePortfolioIntro.mockReset();
    mockUsePortfolioInfo.mockReset();
  });

  it('snapshot test 수정권한이 있을 때', () => {
    mockUsePortfolioIntro.mockImplementation(() => ({
      pfIntro: mockPortfolioIntro,
    }));
    mockUsePortfolioInfo.mockImplementation(() => ({ permissionModify: true }));

    const { asFragment } = render(withAllContext(<PortfolioIntro />));

    expect(screen.getByText('수정')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('snapshot test 수정권한이 없을 때', () => {
    mockUsePortfolioIntro.mockImplementation(() => ({
      pfIntro: mockPortfolioIntro,
    }));
    mockUsePortfolioInfo.mockImplementation(() => ({
      permissionModify: false,
    }));

    const { asFragment } = render(withAllContext(<PortfolioIntro />));

    expect(screen.queryByText('수정')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
