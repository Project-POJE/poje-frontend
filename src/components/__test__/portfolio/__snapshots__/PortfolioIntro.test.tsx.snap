// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioIntro Component snapshot test 수정권한이 없을 때 1`] = `
<DocumentFragment>
  <section
    class="sc-hLBbgP fgZOPR"
  >
    <section
      class="sc-eDvSVe eTrqcI"
    >
      <h3
        class="sc-jSUZER edbCWC"
      >
        제목을 입력해주세요.
      </h3>
      <hr
        class="sc-gKPRtg cPdyCX"
      />
      <p
        class="sc-iBYQkv iqRkiJ"
      >
        내용을 입력해주세요.
      </p>
    </section>
  </section>
</DocumentFragment>
`;

exports[`PortfolioIntro Component snapshot test 수정권한이 있을 때 1`] = `
<DocumentFragment>
  <section
    class="sc-hLBbgP fgZOPR"
  >
    <section
      class="sc-eDvSVe eTrqcI"
    >
      <h3
        class="sc-jSUZER edbCWC"
      >
        제목을 입력해주세요.
      </h3>
      <hr
        class="sc-gKPRtg cPdyCX"
      />
      <p
        class="sc-iBYQkv iqRkiJ"
      >
        내용을 입력해주세요.
      </p>
    </section>
    <article
      class="sc-bcXHqe dUcRao"
    >
      <button
        class="sc-gswNZR sc-dkrFOg ekNcrm"
      >
        수정
      </button>
    </article>
  </section>
</DocumentFragment>
`;
